lib/broadway/producer.ex

```diff
@@ -236,4 +289,110 @@ defmodule Broadway.Producer do
             "struct. In case you're using a standard GenStage producer, please set the " <>
             ":transformer option to transform produced events into message structs"
   end
+
+  defp maybe_rate_limit_and_buffer_messages(state, messages) do
+    if state.rate_limiting && messages != [] do
+      state = update_in(state.rate_limiting.message_buffer, &enqueue_batch(&1, messages))
+      rate_limit_and_buffer_messages(state)
+    else
+      {state, messages}
+    end
+  end
+
+  defp rate_limit_and_buffer_messages(%{rate_limiting: %{state: :closed}} = state) do
+    {state, []}
+  end
+
+  defp rate_limit_and_buffer_messages(%{rate_limiting: %{message_buffer: buffer}} = state) do
```

Comment:
This function does many accesses inside state.rate_limiting, including double `put_in` calls. I think for this one should rather do this in the head:

    %{rate_limiting: rate_limiting} = state

And then in the body:

    %{state: ..., message_buffer: ..., ...} = rate_limiting

With a `put_in(state.rate_limiting, new_rate_limiting)` at the end.